### PR TITLE
fix(core): handle local history append failures

### PR DESCRIPTION
**Pull request by Kinfe123**
## Summary

- keep each user-message history write concurrent with LocalRuntime run startup
- settle the run and history operations together so the history promise no longer floats independently
- preserve run failures as the primary error when both operations fail
- handle the history rejection immediately, including no-run paths where subscriber notification can throw

## Why

`ThreadHistoryAdapter.append()` returns a promise, but the user-message write in `LocalThreadRuntimeCore` was started without being attached to the append task. A rejecting network, IndexedDB, or custom history adapter could therefore reject independently while the model run continued.

The write remains concurrent so persistence latency does not delay generation. The internal append task now settles it and propagates a persistence-only failure through that task. Existing host-facing error policy is unchanged: composer sends consume their send task, while the imperative `ThreadRuntime.append()` wrapper continues to surface non-`MessageNotSentError` failures according to its existing contract.

Regression coverage verifies user-message persistence failure without abandoning the run, no-run persistence failure, and the subscriber-throw edge case.

## Testing

- `@assistant-ui/core`: 111 test files, 1122 tests passed
- `pnpm --filter @assistant-ui/core... build`
- targeted oxlint and oxfmt checks

### CODE REVIEW

File: packages/core/src/runtimes/local/local-thread-runtime-core.ts [hunk 1]
```diff
@@ -335,22 +335,28 @@ export class LocalThreadRuntimeCore
       reason: "unknown",
     });
     this.repository.addOrUpdateMessage(message.parentId, newMessage);
-    this._options.adapters.history?.append({
+    const historyWrite = this._options.adapters.history?.append({
       parentId: message.parentId,
       message: newMessage,
       ...(message.runConfig !== undefined && { runConfig: message.runConfig }),
     });
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/runtimes/local/local-thread-runtime-core.ts:342]: Fixed in 09917638a. The history promise now gets an immediate rejection handler before any subscriber notification, while the original promise is still awaited so its error remains observable. I also added a regression covering a rejected no-run write plus a throwing subscriber.

File: packages/core/src/runtimes/local/local-thread-runtime-core.ts [hunk 2]
```diff
@@ -335,22 +335,28 @@ export class LocalThreadRuntimeCore
...
+          runConfig: message.runConfig ?? {},
+        }),
+        historyWrite,
+      ]);
+      if (runResult.status === "rejected") throw runResult.reason;
+      if (historyResult.status === "rejected") throw historyResult.reason;
     } else {
       this.repository.resetHead(newMessage.id);
       this._notifySubscribers();
+      await historyWrite;
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/runtimes/local/local-thread-runtime-core.ts:360]: This behavior is intentional and is now covered explicitly in 09917638a: a message with startRun: false rejects with the history adapter error, and the model adapter is not called. I also updated the PR description to include the no-run path.

File: packages/core/src/runtimes/local/local-thread-runtime-core.ts [hunk 3]
```diff
@@ -335,22 +335,28 @@ export class LocalThreadRuntimeCore
...
+      const [runResult, historyResult] = await Promise.allSettled([
+        this.startRun({
+          parentId: newMessage.id,
+          sourceId: message.sourceId,
+          runConfig: message.runConfig ?? {},
+        }),
+        historyWrite,
+      ]);
+      if (runResult.status === "rejected") throw runResult.reason;
+      if (historyResult.status === "rejected") throw historyResult.reason;
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/runtimes/local/local-thread-runtime-core.ts:356]: Thanks, this distinction is valid. I updated the PR description to state the narrower behavior precisely: the runtime attaches the history write to its internal append task rather than leaving it floating, while the existing composer and imperative ThreadRuntime error policies remain unchanged. The run still starts concurrently and remains the primary error if both operations fail. The immediate rejection handler also covers exits before the eventual await.

File: packages/core/src/runtimes/local/local-thread-runtime-core.test.ts [hunk 4]
```diff
@@ -143,6 +143,37 @@ describe("LocalThreadRuntimeCore events", () => {
...
+describe("LocalThreadRuntimeCore history persistence", () => {
+  it("surfaces failed user persistence without abandoning the run", async () => {
+    const persistenceError = new Error("history unavailable");
+    const run = vi.fn(async () => ({
+      content: [{ type: "text" as const, text: "done" }],
+    }));
+    const append = vi
+      .fn()
+      .mockRejectedValueOnce(persistenceError)
+      .mockResolvedValueOnce(undefined);
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/core/src/runtimes/local/local-thread-runtime-core.test.ts]: Fixed in 09917638a. The mock now rejects based on item.message.role === "user", so the test continues to assert the intended user-write failure even if write ordering changes.

File: packages/core/src/runtimes/local/local-thread-runtime-core.ts [hunk 5]
```diff
@@ -335,22 +335,29 @@ export class LocalThreadRuntimeCore
...
     } else {
       this.repository.resetHead(newMessage.id);
       this._notifySubscribers();
+      await historyWrite;
```
[the remark this answers is not in the document]
Comment (okisdev) [reply packages/core/src/runtimes/local/local-thread-runtime-core.ts:360]: reject. the no-run path is already covered by `startRun: false` in the second test. `_runAppend` is role-agnostic, so any message that does not start a run settles `historyWrite` the same way. the changeset names user writes because that was the floating promise. the settlement is not user-only.

File: packages/core/src/runtimes/local/local-thread-runtime-core.test.ts [hunk 6]
```diff
@@ -143,6 +143,86 @@ describe("LocalThreadRuntimeCore events", () => {
   });
 });
 
+describe("LocalThreadRuntimeCore history persistence", () => {
+  it("surfaces failed user persistence without abandoning the run", async () => {
+    const persistenceError = new Error("history unavailable");
```
[the remark this answers is not in the document]
Comment (okisdev) [reply packages/core/src/runtimes/local/local-thread-runtime-core.test.ts:148]: reject. run already wins in the `allSettled` order (`throw runResult.reason` first). the existing test already proves the run is not abandoned when history rejects. a dual-reject case would only restating that line.

File: packages/core/src/runtimes/local/local-thread-runtime-core.test.ts [hunk 7]
```diff
@@ -143,6 +143,86 @@ describe("LocalThreadRuntimeCore events", () => {
...
+        },
+      },
+    );
+    thread.subscribe(() => {
+      throw listenerError;
+    });
+
+    await expect(
+      thread.append({ ...userMessage("hello"), startRun: false }),
+    ).rejects.toBe(listenerError);
```
[the remark this answers is not in the document]
Comment (okisdev) [reply packages/core/src/runtimes/local/local-thread-runtime-core.test.ts:221]: not blocking. `rejects.toBe(listenerError)` documents current `_notifySubscribers` behavior. isolating subscriber throws is a different change. this test is here so a throwing subscriber cannot skip `await historyWrite` and leak an unhandled rejection.

File: packages/core/src/runtimes/local/local-thread-runtime-core.test.ts [hunk 8]
```diff
@@ -143,6 +143,86 @@ describe("LocalThreadRuntimeCore events", () => {
   });
 });
 
+describe("LocalThreadRuntimeCore history persistence", () => {
+  it("surfaces failed user persistence without abandoning the run", async () => {
+    const persistenceError = new Error("history unavailable");
+    const run = vi.fn(async () => ({
+      content: [{ type: "text" as const, text: "done" }],
+    }));
+    const append = vi.fn(async (item: ExportedMessageRepositoryItem) => {
```
[the remark this answers is not in the document]
Comment (okisdev) [reply packages/core/src/runtimes/local/local-thread-runtime-core.test.ts:152]: reject. same as the other dual-reject note: run already wins, and the existing case already proves the run still starts. not expanding this PR for another assertion of the same `allSettled` order.

File: packages/core/src/runtimes/local/local-thread-runtime-core.test.ts [hunk 9]
```diff
@@ -143,6 +143,86 @@ describe("LocalThreadRuntimeCore events", () => {
...
+      },
+    );
+    thread.subscribe(() => {
+      throw listenerError;
+    });
+
+    await expect(
+      thread.append({ ...userMessage("hello"), startRun: false }),
+    ).rejects.toBe(listenerError);
+    await flush();
```
[the remark this answers is not in the document]
Comment (okisdev) [reply packages/core/src/runtimes/local/local-thread-runtime-core.test.ts:222]: not blocking. without the `void historyWrite?.catch` the suite already fails on the unhandled `history unavailable` rejection. an explicit `unhandledRejection` listener would only restate that.

**Review by okisdev — APPROVED** [review]
the floating user-history write was real, and settling it next to startRun with allSettled is the right shape. run stays concurrent and still wins if both fail. I ran the persistence suite both ways: with the patch the new cases pass, without it the first two resolve and leak an unhandled `history unavailable` rejection.

the bot notes on a dual-reject case and an explicit unhandledRejection listener are polish, not a missing behavior. composer still swallows non-MessageNotSentError send-task failures, which this PR deliberately left alone.
